pkg/result/processors/sort_results.go

```diff
@@ -0,0 +1,179 @@
+package processors
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/golangci/golangci-lint/pkg/config"
+	"github.com/golangci/golangci-lint/pkg/result"
+)
+
+// Base propose of this functionality to sort results (issues)
+// produced by various linters by analyzing code. We achieving this
+// by sorting results.Issues using processor step, and chain based
+// rules that can compare different properties of the Issues struct.
+
+var _ Processor = (*SortResults)(nil)
+
+type SortResults struct {
+	cmp Comparator
+	cfg *config.Config
+}
+
+func NewSortResults(cfg *config.Config) *SortResults {
+	// For sorting we are comparing (in next order): file names, line numbers,
+	// position, range numbers (if position not available), and finally - giving up.
+	return &SortResults{
+		cmp: ByName{
+			next: ByLine{
+				next: ByColumn{},
+			},
+		},
+		cfg: cfg,
+	}
+}
+
+// Process is performing sorting of the result issues.
+func (sr SortResults) Process(issues []result.Issue) ([]result.Issue, error) {
+	if !sr.cfg.Output.SortResults {
+		return issues, nil
+	}
+
+	sort.Slice(issues, func(i, j int) bool {
+		return sr.cmp.Compare(&issues[i], &issues[j]) == Less
+	})
+
+	return issues, nil
+}
+
+func (sr SortResults) Name() string { return "sort_results" }
+func (sr SortResults) Finish()      {}
+
+type CompareResult int
+
+const (
+	Less CompareResult = iota - 1
+	Equal
+	Greater
+	None
+)
+
+func (c CompareResult) isNeutral() bool {
+	// return true if compare result is incomparable or equal.
+	return c == None || c == Equal
+}
+
+func (c CompareResult) String() string {
+	switch c {
+	case Less:
+		return "Less"
+	case Equal:
+		return "Equal"
+	case Greater:
+		return "Greater"
+	}
+
+	return "None"
+}
+
+// Comparator describe how to implement compare for two "issues" lexicographically
+type Comparator interface {
+	Compare(a, b *result.Issue) CompareResult
+	Next() Comparator
+}
+
+var (
+	_ Comparator = (*ByName)(nil)
+	_ Comparator = (*ByLine)(nil)
+	_ Comparator = (*ByColumn)(nil)
+	_ Comparator = (*ByRange)(nil)
+)
+
+type ByName struct{ next Comparator }
+
+func (cmp ByName) Next() Comparator { return cmp.next }
+func (cmp ByName) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = CompareResult(strings.Compare(a.FilePath(), b.FilePath())); !res.isNeutral() {
+		return res
+	}
+
+	if next := cmp.Next(); next != nil {
+		return next.Compare(a, b)
+	}
+
+	return res
+}
+
+type ByLine struct{ next Comparator }
+
+func (cmp ByLine) Next() Comparator { return cmp.next }
+func (cmp ByLine) Compare(a, b *result.Issue) CompareResult {
+	var res CompareResult
+
+	if res = numericCompare(a.Line(), b.Line()); !res.isNeutral() {
+		return res
+	} else if next := cmp.Next(); res.isNeutral() && next != nil {
```

Comment:
nits: similar to previous comment. The check for res.isNeutral() can be removed like you have done for ByName